File: src/WinMix.Audio/PolicyConfigInterop.cs
using System.Runtime.InteropServices;
using NAudio.CoreAudioApi;

namespace WinMix.Audio;

/// <summary>
/// Undocumented but stable since Windows 7: Core Audio has no public API to change
/// which render device is the system default, so Windows' own Settings app drives
/// this same COM object internally.
///
/// The leading placeholder methods stand in for real ones this app never calls --
/// removing them would shift <see cref="SetDefaultEndpoint"/> to the wrong vtable
/// slot and silently invoke something else instead.
/// </summary>
[ComImport]
[Guid("F8679F50-850A-41CF-9C72-430F290290C8")]
[InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
internal interface IPolicyConfig
{
    void Unused1(); // GetMixFormat
    void Unused2(); // GetDeviceFormat
    void Unused3(); // SetDeviceFormat
    void Unused4(); // GetProcessingPeriod
    void Unused5(); // SetProcessingPeriod
    void Unused6(); // GetShareMode
    void Unused7(); // SetShareMode
    void Unused8(); // GetPropertyValue
    void Unused9(); // SetPropertyValue

    void SetDefaultEndpoint([MarshalAs(UnmanagedType.LPWStr)] string deviceId, Role role);

    void Unused10(); // SetEndpointVisibility
}

[ComImport]
[Guid("870AF99C-171D-4F9E-AF0D-E63DF40C2BC9")]
internal class PolicyConfigClient
{
}

/// <summary>Switches the system's default render device via <see cref="IPolicyConfig"/>.</summary>
internal static class DefaultEndpointSwitcher
{
    /// <summary>
    /// Sets <paramref name="deviceId"/> as the default for every role. Windows'
    /// own device switcher does the same -- otherwise games and notification
    /// sounds (Console), music (Multimedia), and calls (Communications) could
    /// each keep pointing at a different, now-stale device.
    /// </summary>
    public static void SetDefault(string deviceId)
    {
        var policyConfig = (IPolicyConfig)new PolicyConfigClient();
        policyConfig.SetDefaultEndpoint(deviceId, Role.Console);
        policyConfig.SetDefaultEndpoint(deviceId, Role.Multimedia);
        policyConfig.SetDefaultEndpoint(deviceId, Role.Communications);
    }
}

File: WinMix.Audio/src/PolicyConfigInterop.cpp
#include "winmix/audio/PolicyConfigInterop.h"

#include <mmdeviceapi.h>
#include <wrl/client.h>

using Microsoft::WRL::ComPtr;

namespace winmix::audio {

namespace {

// Undocumented (IID F8679F50-850A-41CF-9C72-430F290290C8). The leading
// placeholder methods stand in for real vtable methods (GetMixFormat,
// GetDeviceFormat, ResetDeviceFormat, SetDeviceFormat, GetProcessingPeriod,
// SetProcessingPeriod, GetShareMode, SetShareMode, GetPropertyValue,
// SetPropertyValue) this app never calls -- removing any of them would
// shift SetDefaultEndpoint to the wrong slot and silently invoke a
// different method. A trailing placeholder (SetEndpointVisibility) follows
// the one method actually used, for the same reason.
//
// The original reference source (WinMix.Audio/PolicyConfigInterop.cs) was
// missing ResetDeviceFormat (9 unused methods instead of 10), which put
// SetDefaultEndpoint at the wrong slot -- confirmed by a crash on the very
// first call in an isolated repro (tools/AudioSmokeTest --set-default),
// diagnosed here and cross-checked against the widely-used community
// IPolicyConfig declaration (e.g. the one behind EarTrumpet/SoundVolumeView-
// style tools), which includes it.
MIDL_INTERFACE("F8679F50-850A-41CF-9C72-430F290290C8")
IPolicyConfig : public IUnknown
{
public:
    virtual HRESULT STDMETHODCALLTYPE Unused1() = 0; // GetMixFormat
    virtual HRESULT STDMETHODCALLTYPE Unused2() = 0; // GetDeviceFormat
    virtual HRESULT STDMETHODCALLTYPE Unused3() = 0; // ResetDeviceFormat
    virtual HRESULT STDMETHODCALLTYPE Unused4() = 0; // SetDeviceFormat
    virtual HRESULT STDMETHODCALLTYPE Unused5() = 0; // GetProcessingPeriod
    virtual HRESULT STDMETHODCALLTYPE Unused6() = 0; // SetProcessingPeriod
    virtual HRESULT STDMETHODCALLTYPE Unused7() = 0; // GetShareMode
    virtual HRESULT STDMETHODCALLTYPE Unused8() = 0; // SetShareMode
    virtual HRESULT STDMETHODCALLTYPE Unused9() = 0; // GetPropertyValue
    virtual HRESULT STDMETHODCALLTYPE Unused10() = 0; // SetPropertyValue
    virtual HRESULT STDMETHODCALLTYPE SetDefaultEndpoint(LPCWSTR deviceId, ERole role) = 0;
    virtual HRESULT STDMETHODCALLTYPE Unused11() = 0; // SetEndpointVisibility
};

// CLSID 870AF99C-171D-4F9E-AF0D-E63DF40C2BC9 ("PolicyConfigClient").
constexpr CLSID kPolicyConfigClientClsid = {
    0x870AF99C, 0x171D, 0x4F9E, {0xAF, 0x0D, 0xE6, 0x3D, 0xF4, 0x0C, 0x2B, 0xC9}};

} // namespace

HRESULT DefaultEndpointSwitcher::SetDefault(const std::wstring& deviceId)
{
    ComPtr<IPolicyConfig> policyConfig;
    HRESULT hr = CoCreateInstance(
        kPolicyConfigClientClsid, nullptr, CLSCTX_ALL, IID_PPV_ARGS(&policyConfig));
    if (FAILED(hr))
    {
        return hr;
    }

    // The system default must move games/notifications, music, and calls
    // together, or they would point at different, now-stale devices --
    // unlike AppOutputRouter's per-app pin, this one does include
    // Communications.
    for (const ERole role : {eConsole, eMultimedia, eCommunications})
    {
        hr = policyConfig->SetDefaultEndpoint(deviceId.c_str(), role);
        if (FAILED(hr))
        {
            return hr;
        }
    }

    return S_OK;
}

} // namespace winmix::audio
